Load the data and get the information 

In [1]:
import pandas as pd 


In [6]:
df = pd.read_csv("ecommerce_orders_dirty.csv")


In [7]:
df.head()

,OrderID,CustomerName,Category,Price,Quantity,PaymentMethod,OrderStatus,OrderDate,Rating
0,ORD1396,Sam Rossi,Book,1289.9,3.0,net banking,Cancelled,02/11/2023,3.9
1,ORD1190,Emma Khan,Clothing,341.6 INR,3.0,Debit Card,delivered,2023-12-26,NaN
2,ORD1163,David Garcia,BOOKS,2204.4 INR,3.0,DEBIT CARD,CANCELLED,04/12/2023,3.9
3,ORD1102,Amit Müller,Sports,1105.39,1.0,UPI,DELIVERED,10-07-2023,NaN
4,ORD1362,David Patel,home & kitchen,1808.33 INR,6.0,COD,shipped,19/05/2023,4.8


In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 525 entries, 0 to 524
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   OrderID        525 non-null    str    
 1   CustomerName   490 non-null    str    
 2   Category       496 non-null    str    
 3   Price          509 non-null    str    
 4   Quantity       492 non-null    float64
 5   PaymentMethod  496 non-null    str    
 6   OrderStatus    525 non-null    str    
 7   OrderDate      502 non-null    str    
 8   Rating         484 non-null    float64
dtypes: float64(2), str(7)
memory usage: 37.0 KB


In [9]:
df.describe()


,Quantity,Rating
count,492.000000,484.000000
mean,5.995935,3.043388
std,38.759691,1.478086
min,-3.000000,-1.000000
25%,2.000000,2.000000
50%,3.000000,3.100000
75%,4.000000,4.100000
max,500.000000,10.000000


In [12]:
df.shape

(525, 9)

In [13]:
df.columns

Index(['OrderID', 'CustomerName', 'Category', 'Price', 'Quantity',
       'PaymentMethod', 'OrderStatus', 'OrderDate', 'Rating'],
      dtype='str')

In [14]:
df.dtypes

OrderID              str
CustomerName         str
Category             str
Price                str
Quantity         float64
PaymentMethod        str
OrderStatus          str
OrderDate            str
Rating           float64
dtype: object

Handling the missing values 

In [15]:
df.isnull().sum()

OrderID           0
CustomerName     35
Category         29
Price            16
Quantity         33
PaymentMethod    29
OrderStatus       0
OrderDate        23
Rating           41
dtype: int64

In [30]:
df["Category"] = df["Category"].fillna(df["Category"].mode())

In [17]:
df["PaymentMethod"] = df["PaymentMethod"].fillna("cash")

In [21]:
df["Price"] = df["Price"].astype(str).str.replace(r'[^0-9.]', '', regex=True)
df["Price"] = pd.to_numeric(df["Price"], errors='coerce')

In [22]:
df["Price"] = df["Price"].fillna(df["Price"].mean())

In [23]:
df["Quantity"] = df["Quantity"].fillna(df["Quantity"].mean())

In [24]:
df["Rating"] = df["Rating"].fillna(df["Rating"].mean())

In [26]:
df["OrderDate"] = pd.to_datetime(df["OrderDate"], format="mixed", errors="coerce")

In [27]:
df = df.dropna(subset=["CustomerName", "OrderDate"])

In [31]:
df.info()


<class 'pandas.DataFrame'>
Index: 468 entries, 0 to 524
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   OrderID        468 non-null    str           
 1   CustomerName   468 non-null    str           
 2   Category       468 non-null    str           
 3   Price          468 non-null    float64       
 4   Quantity       468 non-null    float64       
 5   PaymentMethod  468 non-null    str           
 6   OrderStatus    468 non-null    str           
 7   OrderDate      468 non-null    datetime64[us]
 8   Rating         468 non-null    float64       
dtypes: datetime64[us](1), float64(3), str(5)
memory usage: 36.6 KB


In [32]:
df.duplicated().sum()


np.int64(23)

In [33]:
df = df.drop_duplicates()

In [34]:
df = df.drop_duplicates()

In [35]:
df.dtypes

OrderID                     str
CustomerName                str
Category                    str
Price                   float64
Quantity                float64
PaymentMethod               str
OrderStatus                 str
OrderDate        datetime64[us]
Rating                  float64
dtype: object

In [36]:
df["CustomerName"] = df["CustomerName"].str.lower()

In [37]:
df["CustomerName"] = df["CustomerName"].str.strip()

In [38]:
df["CustomerName"] = df["CustomerName"].str.title()

In [39]:
df["Category"] = df["Category"].str.lower()
df["Category"] = df["Category"].str.strip()
df["Category"] = df["Category"].str.title()

In [40]:
df["PaymentMethod"] = df["PaymentMethod"].str.lower()
df["PaymentMethod"] = df["PaymentMethod"].str.strip()
df["PaymentMethod"] = df["PaymentMethod"].str.title()

In [42]:
df["OrderStatus"] = df["OrderStatus"].str.lower()
df["OrderStatus"] = df["OrderStatus"].str.strip()
df["OrderStatus"] = df["OrderStatus"].str.title()

Outliers

In [45]:
Q1 = df["Price"].quantile(0.25)
Q3 = df["Price"].quantile(0.75)
IQR = Q3 - Q1
lower = Q1 - 1.5*IQR
upper = Q3 + 1.5*IQR
outliers = df[
    (df["Price"] < lower) |
    (df["Price"] > upper)
]
df = df[
    (df["Price"] >= lower) &
    (df["Price"] <= upper)]

In [46]:
Q1 = df["Quantity"].quantile(0.25)
Q3 = df["Quantity"].quantile(0.75)
IQR = Q3 - Q1
lower = Q1 - 1.5*IQR
upper = Q3 + 1.5*IQR
outliers = df[
    (df["Quantity"] < lower) |
    (df["Quantity"] > upper)
]
df = df[
    (df["Quantity"] >= lower) &
    (df["Quantity"] <= upper)]

In [47]:
Q1 = df["Rating"].quantile(0.25)
Q3 = df["Rating"].quantile(0.75)
IQR = Q3 - Q1
lower = Q1 - 1.5*IQR
upper = Q3 + 1.5*IQR
outliers = df[
    (df["Rating"] < lower) |
    (df["Rating"] > upper)
]
df = df[
    (df["Rating"] >= lower) &
    (df["Rating"] <= upper)]

In [48]:
from scipy.stats import zscore

z = zscore(df["Price"])

df = df[abs(z) < 3]

In [49]:
from scipy.stats import zscore

z = zscore(df["Quantity"])

df = df[abs(z) < 3]

In [50]:
from scipy.stats import zscore

z = zscore(df["Rating"])

df = df[abs(z) < 3]

In [51]:
df["Price"] = pd.to_numeric(
    df["Price"],
    errors="coerce"
)

In [52]:
df.info()

df.describe()

df.isnull().sum()

df.duplicated().sum()

<class 'pandas.DataFrame'>
Index: 423 entries, 0 to 524
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   OrderID        423 non-null    str           
 1   CustomerName   423 non-null    str           
 2   Category       423 non-null    str           
 3   Price          423 non-null    float64       
 4   Quantity       423 non-null    float64       
 5   PaymentMethod  423 non-null    str           
 6   OrderStatus    423 non-null    str           
 7   OrderDate      423 non-null    datetime64[us]
 8   Rating         423 non-null    float64       
dtypes: datetime64[us](1), float64(3), str(5)
memory usage: 33.0 KB


np.int64(0)

In [ ]:
df.isnull().sum()

df.duplicated().sum()

np.int64(0)

Label encoding 

In [54]:
from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()

df["Category"] = encoder.fit_transform(df["Category"])

one hot encoding 

In [55]:
df = pd.get_dummies(
    df,
    columns=["PaymentMethod", "OrderStatus"],
    drop_first=True
)

In [57]:
df.to_csv(
    "ecommerce_cleaned.csv",
    index=False
)

In [58]:
df.shape

(423, 22)

In [60]:
df.dtypes

OrderID                                      str
CustomerName                                 str
Category                                   int64
Price                                    float64
Quantity                                 float64
OrderDate                         datetime64[us]
Rating                                   float64
PaymentMethod_Cash On Delivery              bool
PaymentMethod_Cc                            bool
PaymentMethod_Cod                           bool
PaymentMethod_Credit Card                   bool
PaymentMethod_Credit_Card                   bool
PaymentMethod_Debit Card                    bool
PaymentMethod_Debit_Card                    bool
PaymentMethod_Net Banking                   bool
PaymentMethod_Netbanking                    bool
PaymentMethod_Upi                           bool
OrderStatus_Cancelled                       bool
OrderStatus_Delivered                       bool
OrderStatus_Pending                         bool
OrderStatus_Returned

In [61]:
df.isnull().sum()

OrderID                           0
CustomerName                      0
Category                          0
Price                             0
Quantity                          0
OrderDate                         0
Rating                            0
PaymentMethod_Cash On Delivery    0
PaymentMethod_Cc                  0
PaymentMethod_Cod                 0
PaymentMethod_Credit Card         0
PaymentMethod_Credit_Card         0
PaymentMethod_Debit Card          0
PaymentMethod_Debit_Card          0
PaymentMethod_Net Banking         0
PaymentMethod_Netbanking          0
PaymentMethod_Upi                 0
OrderStatus_Cancelled             0
OrderStatus_Delivered             0
OrderStatus_Pending               0
OrderStatus_Returned              0
OrderStatus_Shipped               0
dtype: int64

Before cleaning 

Description 
RangeIndex: 525 entries, 0 to 524
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   OrderID        525 non-null    str    
 1   CustomerName   490 non-null    str    
 2   Category       496 non-null    str    
 3   Price          509 non-null    str    
 4   Quantity       492 non-null    float64
 5   PaymentMethod  496 non-null    str    
 6   OrderStatus    525 non-null    str    
 7   OrderDate      502 non-null    str    
 8   Rating         484 non-null    float64
dtypes: float64(2), str(7)
memory usage: 37.0 KB


Shape 
(525, 9)

Type 
OrderID              str
CustomerName         str
Category             str
Price                str
Quantity         float64
PaymentMethod        str
OrderStatus          str
OrderDate            str
Rating           float64
dtype: object

Missing values
OrderID           0
CustomerName     35
Category         29
Price            16
Quantity         33
PaymentMethod    29
OrderStatus       0
OrderDate        23
Rating           41
dtype: int64

After cleaning 

Index: 423 entries, 0 to 524
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   OrderID        423 non-null    str           
 1   CustomerName   423 non-null    str           
 2   Category       423 non-null    str           
 3   Price          423 non-null    float64       
 4   Quantity       423 non-null    float64       
 5   PaymentMethod  423 non-null    str           
 6   OrderStatus    423 non-null    str           
 7   OrderDate      423 non-null    datetime64[us]
 8   Rating         423 non-null    float64       
dtypes: datetime64[us](1), float64(3), str(5)
memory usage: 33.0 KB

Shape 
(423, 22)

Types

OrderID                                      str
CustomerName                                 str
Category                                   int64
Price                                    float64
Quantity                                 float64
OrderDate                         datetime64[us]
Rating                                   float64
PaymentMethod_Cash On Delivery              bool
PaymentMethod_Cc                            bool
PaymentMethod_Cod                           bool
PaymentMethod_Credit Card                   bool
PaymentMethod_Credit_Card                   bool
PaymentMethod_Debit Card                    bool
PaymentMethod_Debit_Card                    bool
PaymentMethod_Net Banking                   bool
PaymentMethod_Netbanking                    bool
PaymentMethod_Upi                           bool
OrderStatus_Cancelled                       bool
OrderStatus_Delivered                       bool
OrderStatus_Pending                         bool
OrderStatus_Returned                        bool
OrderStatus_Shipped                         bool
dtype: object

Missing values

OrderID                           0
CustomerName                      0
Category                          0
Price                             0
Quantity                          0
OrderDate                         0
Rating                            0
PaymentMethod_Cash On Delivery    0
PaymentMethod_Cc                  0
PaymentMethod_Cod                 0
PaymentMethod_Credit Card         0
PaymentMethod_Credit_Card         0
PaymentMethod_Debit Card          0
PaymentMethod_Debit_Card          0
PaymentMethod_Net Banking         0
PaymentMethod_Netbanking          0
PaymentMethod_Upi                 0
OrderStatus_Cancelled             0
OrderStatus_Delivered             0
OrderStatus_Pending               0
OrderStatus_Returned              0
OrderStatus_Shipped               0
dtype: int64


